File: patches/teamwork.cpp
#include <set>
#include "teamwork.h"
#include "genie/dat/DatFile.h"


void makeTechEffectsShared(genie::DatFile *df) {
    for (auto &effect: df->Effects) {
        if (shouldSkipEffect(effect)) {
            continue;
        }
        for (auto &command: effect.EffectCommands) {
            if (command.Type < 8) {
                uint8_t newType = command.Type + 10;
                std::cout << "Modifying command type from " << std::to_string(command.Type) << " to "
                          << std::to_string(newType) << " for effect " << effect.Name << std::endl;
                command.Type = newType;
            }
        }
    }
}

void makeFarmTechsBuffSpeed(genie::DatFile *df) {
    std::vector<int16_t> unitIds = {FARMER_F, FARMER_M, FARM};
    std::vector<int16_t> effectIds = {HORSE_COLLAR, HEAVY_PLOW, CROP_ROTATION};

    for (const auto &effectId: effectIds) {
        for (const auto &unitId: unitIds) {
            auto effectCommand = new genie::EffectCommand();
            effectCommand->Type = 5;  // Attribute Modifier (Multiply)
            effectCommand->A = unitId;
            effectCommand->B = -1;
            effectCommand->C = 13;     // Work Rate
            effectCommand->D = 1.1;    // Amount *
            df->Effects.at(effectId).EffectCommands.push_back(*effectCommand);
            std::cout << "Adding EffectCommand to effect " << effectId << " to multiply work rate of unit " << unitId
                      << " by 1.1" << std::endl;
        }
    }
}

bool shouldSkipEffect(genie::Effect &effect) {
    const std::set<std::string> allowedEffectNames = {
            "Anarchy",
            "Marauders",
            "Nomads",
            "Imperial Nomads",
            "Madrasah",
            "Thalassocracy",
            "Tatars UT 2",
            "Chemistry",
            "Huns Tech Tree",
    };
    const std::set<std::string> forbiddenEffectNames = {
            "Walls HP castle age",
            "Palisade Walls HP feudal age",
    };
    if (allowedEffectNames.find(effect.Name) != allowedEffectNames.end()) {
        return false;
    }
    if (forbiddenEffectNames.find(effect.Name) != forbiddenEffectNames.end()) {
        return true;
    }
    for (auto &command: effect.EffectCommands) {
        if (command.Type == ENABLE_DISABLE_UNIT || command.Type == UPGRADE_UNIT) {
            std::cout << "Skipping effect " << effect.Name << std::endl;
            return true;
        }
    }
    return false;
}

File: mods/teamwork.py
import logging

from genieutils.datfile import DatFile
from genieutils.effect import Effect, EffectCommand

from mods.ids import TYPE_ENABLE_DISABLE_UNIT, TYPE_UPGRADE_UNIT, FARMER_F, FARMER_M, FARM, \
    HORSE_COLLAR, HEAVY_PLOW, CROP_ROTATION, ATTR_WORK_RATE, ATTR_MODIFIER_MULTIPLY

NAME = 'teamwork'

ALLOWED_EFFECTS = [
    "Anarchy",
    "Marauders",
    "Nomads",
    "Imperial Nomads",
    "Madrasah",
    "Thalassocracy",
    "Tatars UT 2",
    "Chemistry",
    "Huns Tech Tree",
]

FORBIDDEN_EFFECTS = [
    "Walls HP castle age",
    "Palisade Walls HP feudal age",
]


def should_skip(effect: Effect):
    if effect.name in ALLOWED_EFFECTS:
        return False
    if effect.name in FORBIDDEN_EFFECTS:
        return True
    for command in effect.effect_commands:
        if command.type == TYPE_ENABLE_DISABLE_UNIT or command.type == TYPE_UPGRADE_UNIT:
            return True
    return False


def make_tech_effects_shared(data: DatFile):
    for effect in data.effects:
        if not should_skip(effect):
            for command in effect.effect_commands:
                if command.type < 8:
                    new_type = command.type + 10
                    logging.info(f'Modifying command type from {command.type} to {new_type} for effect {effect.name}')
                    command.type = new_type


def make_farm_techs_buff_speed(data: DatFile):
    unit_ids = [FARMER_F, FARMER_M, FARM]
    effect_ids = [HORSE_COLLAR, HEAVY_PLOW, CROP_ROTATION]

    for effect_id in effect_ids:
        for unit_id in unit_ids:
            effect_command = EffectCommand(
                type=ATTR_MODIFIER_MULTIPLY,
                a=unit_id,
                b=-1,
                c=ATTR_WORK_RATE,
                d=1.1,
            )
            data.effects[effect_id].effect_commands.append(effect_command)
            logging.info(f'Adding EffectCommand to effect {effect_id} to multiply work rate of unit {unit_id} by 1.1')


def mod(data: DatFile):
    make_tech_effects_shared(data)
    make_farm_techs_buff_speed(data)
